### ISSUE TITLE
Impure nixos on other Linux distro

### ISSUE THREAD

**Issue by holymonson** [NixOS/nixpkgs#110105]
I'm using other Linux distro with nixpkgs, and want to switch to NixOS but not entirely. Compare current implements
- nixpkgs, can only manage userspace packages, not system level services like plasma.
- nixos, tyrannically controlling the whole OS and everything.
- nixos with `isContainer = true;`, boot at `stage-2`, leave bootloader and mount untouched, but still mess /etc e.g. `passwd`.

I'm thinking we may split out a `stage-3`, between container and nixpkgs, to control services only. It should
- leave hardware or filesystem related untouched.
- live alone with other distro same as nixpkgs
- only depends on host's systemd (and kernel)
- don't mess /etc unless we explicitly make config in `configuration.nix`

This could benefit in
- easier to switch to NixOS. Since new user to NixOS is usually an experienced Linux user who are already running their own Linux, it will take much time to deploy a new OS.
- take advantage of other distro, at least in my view, archlinux's AUR provides more convenience than nixpkgs for many standalone or precompiled packages.
- easier to deploy to SBS like raspberry pi because they may have specific well tuned system.

I have done some test but before pushing further, I want to ask maintainers' options or plans, in case it should be called off in advance.

**Comment by veprbl** [comment]
Are you aware of https://github.com/nix-community/home-manager ?

**Comment by holymonson** [comment]
> Are you aware of https://github.com/nix-community/home-manager ?

Sadly home-manager is to control __user specific (non global) packages__, while here we still want nixos to control global packages, but just skip those fundamental skeleton to `/etc`.

**Comment by hmenke** [comment]
@holymonson You can use home-manager to manage `/etc`, see https://gist.github.com/danbst/eb55f9363c0b9151d65271ae6c20f39b

Otherwise maybe something like https://github.com/svanderburg/nix-processmgmt might fit your needs?

**Comment by mjlbach** [comment]
I would agree it would be nice to have something like nix-darwin for FHS compliant linux systems. I often want to run GUI packages on Fedora (installed via nix) or QT and frequently run into issues where wrappers/paths that are assumed on nixos are absent. Some of this is addressed by nixGL, but a lot of it isn't.

**Comment by holymonson** [comment]
@hmenke It' not about managing `/etc` or processes, staging nixos's operations. For example, here is a configuration
```nix
{ config, pkgs, ... }: {
  services.sshd.enable = true;
}
```
nixos will install `openssh` and register `sshd` to enable and start via systemd, this is exactly what I want, neither more nor less.

However, even for the simple configuration above, nixos will run some boot stage __implicitly__ to provide a fundamental Linux, such as to manipulate `passwd` `fstab` `resolv.conf` `ssl/certs`, which are necessary for pure NixOS in bootstrap. But if we are already running on an impure Linux, those bootstrap stages are redundant and even harmful as they may likely damage to the base Linux.

So here we propose, to add an option to skip those bootstrap stages, let nixos could run on an already bootstraped and prepared Linux, like a bare container.

**Comment by holymonson** [comment]
gentle ping @vcunat @danbst @adisbladis @zowoq @Infinisil , since this may be sort of container-related.

**Comment by hmenke** [comment]
There is now https://github.com/numtide/system-manager but it's still in a very early stage.
